### PR TITLE
ISPN-13500 CLI install

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-13500

Adds a dependency on `org.apache.commons:commons-compress`

### CODE REVIEW

**Comment by karesti** [comment]
@tristantarrant I will test this with quarkus

**Comment by tristantarrant** [comment]
It needs the image patch too

File: cli/src/main/java/org/infinispan/cli/download/Artifact.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package org.infinispan.cli.download;
```
Comment (ryanemerson) [inline cli/src/main/java/org/infinispan/cli/download/Artifact.java:1]: Nitpick, but wouldn't `org.infinispan.cli.aritfacts` be more appropriate? Not all artfacts require a download.

File: cli/src/main/java/org/infinispan/cli/util/Utils.java [hunk 2]
```diff
@@ -1,14 +1,87 @@
...
+         return Util.toHexString(digest.digest());
+      } catch (NoSuchFileException e) {
+         return null;
+      } catch (Exception e) {
+         throw new RuntimeException(e);
+      }
+   }
+
+   public static Path downloadFile(URL url, Path dest, boolean verbose) throws IOException {
+      if (Files.exists(dest)) {
```
Comment (ryanemerson) [inline cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: Shouldn't we take into account the artifact's sha here?

For example, if file `example.tgz` is downloaded but corrupt. The initial artifact install will fail due to the wrong sha, however a subsequent attempt will result in `Files.exists` passing and the non-corrupt artifact will never be downloaded. Subsequently all install attempts will fail until the user manually clears the tmp cache.

Instead, if the file exists in the cache with the wrong sha, then the file should be re-downloaded and the old cached value discarded.

**Review by ryanemerson — CHANGES_REQUESTED** [review]

**Review by oraNod — APPROVED** [review]
@tristantarrant This is a great usability improvement. I've pushed a commit with doc suggestions.

File: cli/src/main/java/org/infinispan/cli/commands/Install.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+   @Override
+   public boolean isHelp() {
+      return help;
+   }
+
+   @Override
+   public CommandResult exec(ContextAwareCommandInvocation invocation) throws CommandException {
+      try {
+         StandardCopyOption options[] = overwrite ? new StandardCopyOption[]{StandardCopyOption.REPLACE_EXISTING} : new StandardCopyOption[]{};
+         MavenSettings.getSettings(mavenSettings != null ? mavenSettings.getFile().toPath() : null);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/commands/Install.java:70]: It is done to initialize the singleton.

File: cli/src/main/java/org/infinispan/cli/commands/Install.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.infinispan.cli.download.Artifact;
+import org.infinispan.cli.download.MavenSettings;
+import org.infinispan.cli.impl.ContextAwareCommandInvocation;
+import org.infinispan.cli.logging.Messages;
+import org.kohsuke.MetaInfServices;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 10.0
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/commands/Install.java]: Done

File: cli/src/main/java/org/infinispan/cli/logging/Messages.java [hunk 5]
```diff
@@ -234,4 +234,13 @@
...
    @Message("A username must be specified")
    IllegalArgumentException usernameRequired();
+
+   @Message("Checksum for '%s' does not match. Supplied: %s Actual: %s")
+   SecurityException checksumFailed(String path, String checksum, String computed);
+
+   @Message("Checksum for '%s' verified")
+   String checksumVerified(String path);
+
+   @Message("Artifac '%s' not found")
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/logging/Messages.java]: Done

File: cli/src/main/java/org/infinispan/cli/commands/Install.java [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+   @Override
+   public boolean isHelp() {
+      return help;
+   }
+
+   @Override
+   public CommandResult exec(ContextAwareCommandInvocation invocation) throws CommandException {
+      try {
+         StandardCopyOption options[] = overwrite ? new StandardCopyOption[]{StandardCopyOption.REPLACE_EXISTING} : new StandardCopyOption[]{};
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/commands/Install.java]: Done

File: cli/src/main/java/org/infinispan/cli/download/Artifact.java (hunk 1 above)
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/download/Artifact.java]: Done

File: cli/src/main/java/org/infinispan/cli/util/Utils.java (hunk 2 above)
Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: I have added a new `-f/--force` option which allows you to force downloads, deleting cached artifacts. Obviously it does not work for local artifacts

**Comment by tristantarrant** [comment]
I have integrated Don's commit too

File: cli/src/main/java/org/infinispan/cli/util/Utils.java (hunk 2 above)
Comment (ryanemerson) [reply cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: So what happens if the operator runs into the scenario I describe above? Do we always us `-f`, in which case the caching is never used?

Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: The alternative is to implement retries

Comment (tristantarrant) [reply cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: Also, the Maven stuff I took from WildFly modules doesn't even download the sha's from the remote repos.

Comment (ryanemerson) [reply cli/src/main/java/org/infinispan/cli/util/Utils.java:52]: > The alternative is to implement retries

That works for me as that's the current approach taken by the operator. Ideally such a mechanism should exist in the CLI or image, so the operator can just provide the necessary dependency/patch args to the container.

**Comment by tristantarrant** [comment]
I have fixed the issue with installation from the image https://github.com/infinispan/infinispan-images/pull/157#discussion_r754352016

**Comment by tristantarrant** [comment]
I've added a `-r/--retries=num` option. If a cached/downloaded artifact fails the checksum verification, it is discarded and redownloaded up to num times.

**Comment by tristantarrant** [comment]
Added a `--clean` option which cleans the `server/lib` directory

**Review by ryanemerson — APPROVED** [review]

**Comment by tristantarrant** [comment]
I've made a small change so that the maven `settings.xml` can also be referenced as a URL and it will be downloaded, e.g.
```
./bin/cli.sh install --maven-settings=https://raw.githubusercontent.com/infinispan/infinispan/main/maven-settings.xml ...
```

**Review by karesti — APPROVED** [review]

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**